k8s/store.go

```diff
@@ -0,0 +1,156 @@
+package k8s
+
+import (
+	"errors"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/api/extensions/v1beta1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+)
+
+// Time to handle multiple updates occurring in a short time period, such as at startup where
+// each existing endpoint / ingress produces a single update.
+const bufferedWatcherDuration = time.Millisecond * 50
+
+// Store encapsulates the creation of the shared informer for each watched resource
+type Store interface {
+	// GetOrCreateNamespaceSource create a namespace informer and returns its corresponding store and event handler
+	GetOrCreateNamespaceSource() (*WatchedResource, error)
+	// GetOrCreateIngressSource create an ingress informer and returns its corresponding store and event handler
+	GetOrCreateIngressSource() (*WatchedResource, error)
+	// GetOrCreateServiceSource create a service informer and returns its corresponding store and event handler
+	GetOrCreateServiceSource() (*WatchedResource, error)
+}
+
+type lazyLoadedStore struct {
+	clientset                *kubernetes.Clientset
+	stopCh                   chan struct{}
+	resyncPeriod             time.Duration
+	informerFactory          informerFactory
+	eventHandlerFactory      eventHandlerFactory
+	namespaceWatchedResource *WatchedResource
+	ingressWatchedResource   *WatchedResource
+	serviceWatchedResource   *WatchedResource
+}
+
+// WatchedResource exposes a store for the resource to allow interrogating the current state via get/list operations
+// and an watcher so clients can subscribe to update notifications
+type WatchedResource struct {
```

Comment:
Just a minor quibble with the naming: when I first read `WatchedResource` elsewhere in the code, I thought it was referring to a specific individual named object, e.g. a single ingress.

Reply:
I'm open to suggestion :)

Reply:
`WatchedStore`?